src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs

```diff
@@ -387,6 +410,9 @@ internal DocumentBranch(SourceText text, DocumentState state)
             }
         }
 
+        // This number is arbitrary. It is intended to keep the chain of incremental parses relatively small
+        private const int MAX_INCREMENTAL_PARSE_DEPTH = 32;
```

Comment:
I'm quite leery of checking this in without have decent perf testing of it in advance.